src/components/Widget/inputs.tsx

```diff
@@ -20,30 +20,39 @@ function includesDefaultToken(tokens: SwapTokens) {
  * Enforces that token is a part of the returned value.
  */
 export function useSyncWidgetInputs({
-  token,
-  defaultField,
-  onTokenChange,
+  inputToken,
+  outputToken,
+  defaultToken,
+  onDefaultTokenChange,
 }: {
-  token?: Currency
-  defaultField: Field
-
-  onTokenChange?: (token: Currency) => void
+  inputToken?: Currency
+  outputToken?: Currency
+  defaultToken?: Currency
+  onDefaultTokenChange?: (token: Currency) => void
 }) {
   const trace = useTrace({ section: InterfaceSectionName.WIDGET })
 
   const [type, setType] = useState<SwapValue['type']>(TradeType.EXACT_INPUT)
   const [amount, setAmount] = useState<SwapValue['amount']>(EMPTY_AMOUNT)
-  const [tokens, setTokens] = useState<SwapTokens>({ [defaultField]: token, default: token })
+  const [tokens, setTokens] = useState<SwapTokens>({
+    [Field.INPUT]: inputToken,
+    [Field.OUTPUT]: outputToken,
+    default: defaultToken,
+  })
 
   useEffect(() => {
-    setTokens((tokens) => {
-      const update = { ...tokens, default: token }
-      if (!includesDefaultToken(update)) {
-        return { [defaultField]: update.default, default: update.default }
-      }
-      return update
-    })
-  }, [defaultField, token])
+    if (!tokens[Field.INPUT] && !tokens[Field.OUTPUT]) {
+      setTokens((tokens) => {
+        const update = {
+          ...tokens,
+          [Field.INPUT]: inputToken ?? tokens[Field.INPUT],
+          [Field.OUTPUT]: outputToken ?? tokens[Field.OUTPUT] ?? defaultToken,
+          default: defaultToken,
+        }
+        return update
+      })
+    }
+  }, [defaultToken, inputToken, outputToken, tokens])
```

Comment:
Note that this introduces a delay (because `useEffect` will run after a render with the deps-array values), so it won't look good for controlled values (like `inputToken`/`outputToken`).

Are these actually controlled? Or are they just here for initialization?

I _think_ they're just here for initialization, in which case could they share a prop with `defaultToken`? eg

    defaultTokens: Currency | { input?: Currency, output?: Currency }

Reply:
updated to be one single prop